### PR TITLE
[ENH] Python implementation of select Landtrendr convenience functions

### PR DESCRIPTION
**Pull request by crich011**
This PR is an attempt at a python implementation of some convenience functions from the LT-GEE JavaScript package and some additional JS functions we had previously developed to work with those outputs.

### CODE REVIEW

**Comment by crich011** [comment]
Comparing outputs of these functions, it appears that the functions in `postprocess` produce the same results as their JS equivalents, while the functions in `preprocess` differ slightly, by something like O(0.1%). This agreement is close enough that a good candidate for the source of the disagreement is the selection of the medoid image. It might be worth checking different distance metrics, like ones which use only red, NIR, and SWIR2.

**Review by gersh — COMMENTED** [review]
I think there may off-by-one discrepancy in how the dates are handled in this code from the javascript. Can we test this?

File: earth-engine/landtrendr/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    end_date = ee.Date.fromYMD(ee.Number(year), int(end_m), int(end_d)).advance(
+        1, "day"
+    )
+    filtered_collection = collection.filterDate(start_date, end_date)
+
+    # At this point in LT-GEE implementation, ee.Algorithms.If is used to check
+    # number of input images. However, checking using .gt() appears to not work
+    # with Algorithms.If since it returns a numeric value. For now, punt on
+    # figuring this out and just assume we have a non-empty collection at this point.
+    # https://github.com/eMapR/LT-GEE/blob/b0e92a0c198bdd1a794e1e9b8f4db8fc7fa06054/scripts/python/lt_gee_bap_test.py#L104
```
[the remark this answers is not in the document]
Comment (crich011) [reply earth-engine/landtrendr/preprocess.py:84]: I managed to get the coalescing with an empty collection bit to work, so I'll remove the comment and push the update.

File: earth-engine/landtrendr/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+
+    def _euclidean_distance(image: ee.Image) -> ee.Image:
+        distance = ee.Image(image).subtract(median).pow(ee.Image.constant(2))
+        return distance.reduce("sum").addBands(image)
+
+    distance_from_median = filtered_collection.map(_euclidean_distance)
+    return (
+        ee.ImageCollection(distance_from_median)
+        .reduce(ee.Reducer.min(7))
+        .set("system:time_start", ee.Date.fromYMD(year, 1, 1).millis())
```
[the remark this answers is not in the document]
Comment (crich011) [reply earth-engine/landtrendr/preprocess.py:96]: I went ahead and changed this to be September 1st, in agreement with the JS version, but it shouldn't affect the results.

File: earth-engine/landtrendr/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+    ee.Image
+      An ee.Image containing the pixel-wise medoid for the year in question.
+    """
+    start_m, start_d = start_day.split("-")
+    end_m, end_d = end_day.split("-")
+
+    start_date = ee.Date.fromYMD(ee.Number(year), int(start_m), int(start_d))
+    end_date = ee.Date.fromYMD(ee.Number(year), int(end_m), int(end_d)).advance(
+        1, "day"
+    )
```
[the remark this answers is not in the document]
Comment (crich011) [reply earth-engine/landtrendr/preprocess.py:77]: The `advance` bit is to make the window inclusive, as opposed to exclusive. Changing these to make it exclusive doesn't appear to affect the result (in particular, does not explain the difference between this version and the JS version).

File: earth-engine/landtrendr/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+    end_day: str
+      The last day, inclusive, to consider in calculating the medoid. Formatted as
+      'mm-dd'.
+
+    Returns
+    -------
+    ee.ImageCollection
+      A collection of yearly medoid images for the provided collection.
+    """
+
```
[the remark this answers is not in the document]
Comment (crich011) [reply earth-engine/landtrendr/preprocess.py:121]: Ah, that's a good point; we will need this for processing any non-tropical southern hemisphere areas, so I should take the trouble of adding it.

**Comment by crich011** [comment]
This latest commit addresses all but the bit about handling southern hemisphere summers -- unfortunately, inclusive vs. exclusive for endpoint dates doesn't explain discrepancies with the JS numbers.

**Comment by crich011** [comment]
Given our discussion in #tech_dev, I am going to go ahead and merge this and start working on porting it over to the [Earthshot](https://github.com/earthshot-labs/earthshot) repo.

**Pull request merged**